### PR TITLE
Parse genes partial fusion pairs

### LINKED ISSUE #4286: Gene filter for fusion vcf should work for fusion_gene names
- Gene filter should work for fusion_gene names when fusions not having gene objects associated

### LINKED ISSUE #4507: Fusion specific fields missing from variantS export
It would be nifty to export things like exon number, orientation, maybe split read counts etc, and `fusion_genes` for fallback when they do not correspond to plain HGNC genes that we have loaded.

### PR DESCRIPTION
**Pull request by dnil**
This PR adds a functionality or fixes a bug.

- Partially fix #4286:
  - Fix bug where genes were discarded during load parsing if one of the fusion partners was lacking hgnc_id. These genes constitute the bulk of the issue, and become searchable etc upon case (re)upload.

With the demo case:

Screenshot 2024-03-14 at 14 44 40

Screenshot 2024-03-14 at 14 47 19

- Fix #4507 - added fusion specific fields missing from variantS export. This helps a bit, as the symbols missing a gene are now part of the export.

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by dnil** [comment]
Ok, testing the front end part for now on [stage](https://scout-stage.scilifelab.se/cust000/ACC8848A2-75M-FUSION/fusion/variants?variant_type=clinical). So far, so good I guess.
Before: 
Screenshot 2024-03-14 at 15 00 18

After:
Screenshot 2024-03-14 at 15 15 47

**Comment by dnil** [comment]
Mm, the loading bug for the fusion partner of a non-gene is fixed, which does quite a bit on the stage case at least, but as evident when searching, the front end symbol/alias lookup part will not allow searches on its own. Try e.g. [POLH-AS1](https://scout-stage.scilifelab.se/genes/40459) on the stage case: https://scout-stage.scilifelab.se/cust000/ACC8848A2-75M-FUSION/fusion/variants

**Comment by dnil** [comment]
While not exactly in the original issue, it would be nice to 
- use the alias lookup for fusion_genes also for export
- perhaps store back added gene info from alias lookup to the variant in db, to allow searching etc.

**Comment by dnil** [comment]
Sorry for the bulky PR, but the issues were connected, and fixing one thing helped resolve issues with the other.. 

Fusion variantS export now includes some fusion specific fields:
Screenshot 2024-03-15 at 14 28 56

File: scout/constants/variants_export.py [hunk 1]
```diff
@@ -14,6 +14,16 @@
 ]
```
Comment (northwestwitch) [inline scout/constants/variants_export.py:15]: Not your fault, but have you seen this?

image

🤣 🤣 🤣

**Comment by dnil** [comment]
I would prefer not to refactor those apparently complex functions, but it is kind of trivial to break out different sub functions for the different categories if need be.

File: scout/constants/variants_export.py (hunk 1 above)
Comment (dnil) [reply scout/constants/variants_export.py:15]: Yes... Im part guilty now as well, but it's mostly because of a very looong sequence of "if this value exists, do that". While not so DRY perhaps it's also not really complex. We could break it up in one function per variant category to "fool" the counter, but I dunno..

Comment (northwestwitch) [reply scout/constants/variants_export.py:15]: Let's do it in a dedicated PR!

File: scout/constants/variants_export.py [hunk 2]
```diff
@@ -14,6 +14,16 @@
...
 
 CANCER_EXPORT_HEADER = EXPORT_HEADER + ["VAF TUMOR", "VAF NORMAL", "COSMIC ID"]
+FUSION_EXPORT_HEADER = EXPORT_HEADER + [
```
Comment (northwestwitch) [inline scout/constants/variants_export.py:17]: I would have preferred to handle the export of fusion variants in a separate PR, to be reviewed after merging this one (because I understand this is needed to create content for the export)

File: scout/server/blueprints/variants/controllers.py [hunk 3]
```diff
@@ -1108,7 +1182,35 @@ def variant_export_lines(store, case_obj, variants_query):
         else:
             variant_line.append("N/A")
 
-        if case_obj.get("track") == "cancer":
+        if category == "fusion":
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1185]: This code adds quite a lot of complexity to an already complex function and it would be much better to move outside.

Complexity of main branch with radon:

image


Complexity of this branch:

image

File: scout/server/blueprints/variants/controllers.py [hunk 4]
```diff
@@ -331,20 +333,94 @@ def fusion_variants(
...
+    Also, if we do not have any kind of symbols for the fusion genes, there is not much we can do.
+    If the fusion gene apparent symbol is not already among the variant hgnc symbols, attempt to find the gene using
+    first its exact symbol, then the special case of HGNC locus group symbols, ending in `@` and not kept as aliases,
+    then lastly aliases.
+
+    Update variant_obj with the new info.
+    """
+
+    variant_genes = variant_obj.get("genes") or []
+    if len(variant_genes) >= 2:
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:368]: What if the gene "a" is actually 2 genes and the gene "b" is missing? Can the situation happen?

File: scout/server/blueprints/variants/controllers.py [hunk 5]
```diff
@@ -331,20 +333,94 @@ def fusion_variants(
...
+        return
+
+    variant_hgnc_ids = variant_obj.get("hgnc_ids") or []
+    variant_hgnc_symbols = variant_obj.get("hgnc_symbols") or []
+
+    fusion_genes = variant_obj.get("fusion_genes") or []
+    if not fusion_genes:
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:375]: Perhaps put this on top of the function?

File: scout/server/blueprints/variants/controllers.py [hunk 6]
```diff
@@ -331,20 +333,94 @@ def fusion_variants(
...
+        return
+
+    result_genes = []
+    remaining_fusion_gene_symbols = []
+    for fusion_gene_symbol in fusion_genes:
+        if fusion_gene_symbol in variant_hgnc_symbols:
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:381]: It is quite hard to follow what is going on in this function, also thinking about future maintenance. Can some of this stuff be moved to the parsing part instead? So it doesn't have to be on the fly every time one loads the variantS page

File: scout/server/blueprints/variants/controllers.py [hunk 7]
```diff
@@ -331,20 +333,94 @@ def fusion_variants(
...
+        return
+
+    variant_hgnc_ids = variant_obj.get("hgnc_ids") or []
+    variant_hgnc_symbols = variant_obj.get("hgnc_symbols") or []
+
+    fusion_genes = variant_obj.get("fusion_genes") or []
+    if not fusion_genes:
+        return
+
+    result_genes = []
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:378]: Type hints could help understanding better what these variables contain. Add to the entire function

File: scout/server/blueprints/variants/controllers.py [hunk 8]
```diff
@@ -1024,21 +1100,17 @@ def generate(header, lines):
...
-
     Returns:
         an HTTP response containing a csv file
     """
-    document_header = variants_export_header(case_obj)
-    export_lines = []
+    document_header = variants_export_header(case_obj, category)
     export_lines = variant_export_lines(
-        store, case_obj, variant_objs.limit(EXPORTED_VARIANTS_LIMIT)
+        store, case_obj, variant_objs.limit(EXPORTED_VARIANTS_LIMIT), category
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1113]: Yes, category might really be the key to simplify the export. But I would keep that code outside this PR

File: scout/server/blueprints/variants/controllers.py [hunk 9]
```diff
@@ -1084,18 +1158,18 @@ def variant_export_lines(store, case_obj, variants_query):
         variant_line.append("_".join([str(position), change]))
 
         # gather gene info:
-        gene_list = variant.get("genes")  # this is a list of gene objects
+        gene_list = variant.get("genes", [])  # this is a list of gene objects
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/controllers.py:1161]: 👍🏻

**Review by northwestwitch — COMMENTED** [review]
Just an initial impression: I appreciate the fix of 3 things but I think the code, the review process and the future maintenance would be much better if you separate the variant export part from the missing fusion genes which is the purpose of this PR. 

That way you could dedicate more attention to include the missing fields perhaps it a more stylish way that won't add so much complexity to an already complex function?

File: scout/constants/variants_export.py (hunk 1 above)
Comment (northwestwitch) [reply scout/constants/variants_export.py:15]: By reading this PR I've realised that having a dedicated PR for the export would give that code more justice.. Sorry!

File: scout/constants/variants_export.py (hunk 2 above)
Comment (dnil) [reply scout/constants/variants_export.py:17]: Ok, let's do it like that then. It is more work, and doing it this way was good for uncovering missing values.

File: scout/server/blueprints/variants/controllers.py (hunk 3 above)
Comment (dnil) [reply scout/server/blueprints/variants/controllers.py:1185]: Right, I would still argue it is not complexity, rather something like affine crystallinity or redundancy, but I do agree it should be fixed.

File: scout/server/blueprints/variants/controllers.py (hunk 6 above)
Comment (dnil) [reply scout/server/blueprints/variants/controllers.py:379]: Sure, but then we loose the only large benefit with this function, which is fixing old cases without reload. Perhaps it is cleaner just to remove it and force a reload. Searches would still not be fully functional without upserting the fixed objects to db.

File: scout/server/blueprints/variants/controllers.py (hunk 4 above)
Comment (dnil) [reply scout/server/blueprints/variants/controllers.py:366]: yes, we should check for the same len as the fusion genes, if the latter exists!

**Comment by dnil** [comment]
Streamlined into a PR to bug fix #4503. Made a separate PR for fusion export #4530. Removed the controller side fusion gene symbol lookups. The latter can be revisited when fixing e.g. #4505 or done in the pipeline by extending/updating the annotation set so that more genes have current gene symbols or hgnc_id.

File: scout/parse/variant/variant.py [hunk 10]
```diff
@@ -271,7 +271,6 @@ def get_genmod_key(case):
     """
     case_id = case["_id"]
     if "-" in case_id:
-        LOG.debug("iCase ID contains '-'. Using display_name as case _id.")
```
Comment (northwestwitch) [inline scout/parse/variant/variant.py:274]: 👍🏻

File: scout/server/blueprints/variants/templates/variants/fusion-variants.html [hunk 11]
```diff
@@ -127,7 +127,7 @@
       {% if variant.genes %}
         {% for gene in variant.genes %}
           {% for transcript in gene.transcripts %}
-            <div>{{ transcript.exon|int }} </div>
+            <div>{% if "exon" in transcript %}{{ transcript.exon|int }}{% endif %} </div>
```
Comment (northwestwitch) [inline scout/server/blueprints/variants/templates/variants/fusion-variants.html:130]: 👍🏻

**Review by northwestwitch — APPROVED** [review]
Really nice fix to the missing genes problem! 💯 
And thanks for simplifying the code, I believe it will be easier to maintain in the long run

**Comment by dnil** [comment]
Re-tested on stage. Looks good, and even though I miss the link out to genes found by alias or post-load lookup for old bugged cases, they were a bit of a false friend without permanent storing.
